File: leetcode/queue_stacks.cpp
#include <stdexcept>
#include <vector>

// number: 232
// title: Implement Queue using Stacks
// url: https://leetcode.com/problems/implement-queue-using-stacks/
// section: 
// difficulty: easy
// tags: stack, design, queue, grind 75

// constraints
// 1 <= x <= 9
// At most 100 calls will be made to push, pop, peek, and empty.
// All the calls to pop and peek are valid.

// complexity
// run-time: see below
// space: O(n)
template <typename T>
class MyQueue {
public:
    // run-time: O(1)
    void push(const T& x) {
        d_back.push_back(x);
    }

    // run-time: O(1) amortized, O(n) worst-case
    T pop() {
        if (empty()) {
            throw std::length_error("empty queue");
        }

        if (d_front.empty() && !back2front()) {
            throw std::runtime_error("failed to move back to front");
        }

        T value = d_front.back();
        d_front.pop_back();
        return value;
    }

    // run-time: O(1) amortized, O(n) worst-case
    T peek() {
        if (empty()) {
            throw std::length_error("empty queue");
        }

        if (d_front.empty() && !back2front()) {
            throw std::runtime_error("failed to move back to front");
        }

        return d_front.back();
    }

    // run-time: O(1)
    bool empty() {
        return d_front.empty() and d_back.empty();
    }

private:
    // run-time: O(n)
    bool back2front() {
        if (empty()) {
            return false;
        }

        while (!d_back.empty()) {
            d_front.push_back(d_back.back());
            d_back.pop_back();
        }

        return !d_front.empty();
    }
    std::vector<T> d_front;
    std::vector<T> d_back;
};

MyQueue* obj = new MyQueue();
obj->push(x);
int param_2 = obj->pop();
int param_3 = obj->peek();
bool param_4 = obj->empty();

File: leetcode/queue_stacks.py
#!/usr/bin/env python3

import sys
import unittest

# number: 232
# title: Implement Queue using Stacks
# url: https://leetcode.com/problems/implement-queue-using-stacks/
# section: 
# difficulty: easy
# tags: stack, design, queue

# constraints
# 1 <= x <= 9
# At most 100 calls will be made to push, pop, peek, and empty.
# All the calls to pop and peek are valid.

# solution: 2 stacks/lists
# complexity
# run-time: see below
# space: O(n)
class MyQueue:

    def __init__(self):
        self.__front = []
        self.__back = []

    # run-time: O(1)
    def push(self, x: int) -> None:
        # always push to back stack
        self.__back.append(x)

    # run-time: O(1) amortized, O(n) worst-case
    def pop(self) -> int:
        if self.empty():
            raise ValueError("empty queue") 

        # if front is empty, transfer to front
        if not self.__front and not self.__back2front():
            raise ValueError("failed moving from back to front")

        # pop from front stack
        return self.__front.pop()

    # run-time: O(1) amortized, O(n) worst-case
    def peek(self) -> int:
        if self.empty():
            raise ValueError("empty queue")

        if not self.__front and not self.__back2front():
            raise ValueError("failed moving from back to front")

        return self.__front[-1]

    # run-time: O(n)
    def __back2front(self) -> bool:
        if self.empty():
            return False

        while self.__back:
            self.__front.append(self.__back.pop())

        return True

    # run-time: O(1)
    def empty(self) -> bool:
        return not len(self.__front) and not len(self.__back)

# TODO: worst-case O(1)?

class TestMyQueue(unittest.TestCase):
    def test(self):
        q = MyQueue()
        q.push(1)                       # [1]
        q.push(2)                       # [1,2]
        self.assertEqual(q.peek(), 1)
        self.assertEqual(q.pop(), 1)    # [2]
        self.assertFalse(q.empty())

if __name__ == '__main__':
    sys.exit(unittest.main())
